engine/src/main/java/org/hibernate/search/backend/IndexWorkVisitor.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend;
+
+/**
+ * Contract for visitors of {@link LuceneWork} types.
+ * <p>
+ * Implementations can receive a specific input parameter and return a specific output type. Use {@link Void} as
+ * parameterization for each of the two not needed.
+ *
+ * @author Sanne Grinovero
+ * @author Gunnar Morling
+ * @param <P> Context parameter type expected by a specific visitor
+ * @param <R> Return type provided by a specific visitor
+ */
+public interface IndexWorkVisitor<P, R> {
```

Comment:
I replied to it in https://github.com/gunnarmorling/hibernate-search/commit/4f64f6943cbe06629a736c1647386b456f71323c#commitcomment-10700416:

> [The parameter] is needed to pass in visitor-specific context where required. The FlushStrategySelector uses it, but also one implementation in ISPN can take advantage of it. This allows to have the actual work done in the visitor itself rather than returning a delegate.
>
> I only did not do that for the other existing implementations in order to avoid allocation of a context parameter object (which would be needed since more than one contextual object need to be passed in). But the design should allow for this in cases where a) there only is a single context object needed anyways (so no further allocation required) or b) we don't mind the allocation of such wrapper for passing in several context parameters.

It enables exactly what you seem to be after: have visitors which do the actual work. Whether that's not done in specific cases due to performance considerations should not pose a limitation to the design in general IMO.